Restore compatibility with Standard Library modules on OCaml 4.x

dra27 opened the pull request:
Use the Standard Library modules for OCaml 4.x, but override 4.14's to eliminate the deprecation warning.

dra27 committed:
Only install modules for 4.14+

Use the Standard Library modules for OCaml 4.x, but override 4.14's to
eliminate the deprecation warning.

nojb reviewed (dismissed):
LGTM

Octachron commented:
A potential problem with the special casing of 4.14 is that it would be the only version where `Stdlib.Stream.t` and `Stream.t` are valid but incompatible types.

dra27 commented:
Very good point! I’ll push a revision shortly

dra27 committed:
Re-export 4.14 modules without deprecation

dra27 commented:
Not the prettiest workaround, but I've amended the `dune` so that:
- For 4.13 and earlier, it remains an empty library
- For 5.0 and later, the files in src/ are used
- For 4.14, one-liners are generated re-exporting the Standard Library modules, but without the deprecations

nojb commented on dune:
```diff
@@ -0,0 +1,51 @@
+  (target %s.mli)
+  (action (with-stdout-to %%{target}
+    (echo "[@@@ocaml.warning \"-3\"] include module type of %s")))
+  (enabled_if (< %%{ocaml_version} 5.0)))
```
Do we need this `.mli`? I am always a bit suspicious of `module type of`...

dra27 replied:
Argh - I think you're right... it was supposed to have been `include module type of struct include %s end`, although it is working in this case. @Octachron - is there any problem with just having the header generated from `include Stream` in the `.ml` file?

Octachron replied:
Including `Stream` creates a slight documentation issue due to the presence of hidden functions in the interface of `Stream`. I am not sure if there is alternative (outside of conditionally adding the type equation for 4.14 in the compatibility mli file?)

nojb reviewed (approved):
LGTM

xavierleroy commented:
Apologies for a late and basic question, but: what's the motivation? what problem is this PR addressing?

dra27 commented:
#4 - at present camlp-streams doesn’t work properly on 4.04-4.06. I just need to push a revised rule for the .mli file.

xavierleroy commented:
Ah, thanks for the reminder.  We do need to make the package empty for OCaml < 4.07, indeed.  For 4.07-4.13, I'm not sure what the best course of action is.

dra27 commented:
I think it’s best to do nothing all the way up to 4.13 - 4.04-4.06 have an immediately-visible linking problem, but there’s a type equality problem for 4.07 onwards too, IIRC

dra27 committed:
Strengthened types

dra27 committed:
Add a linking and type equality tests

dra27 commented:
I added two tests: one explicitly tests linking an executable which camlp-streams and a library which _doesn't_ use camlp-streams (that's the original issue reported in #4) and another which fails of `Stream.t` from camlp-streams isn't equal to `Stdlib.Stream.t`, which has also been coming up.

The tests fail as expected (the link test fails on 4.02-4.06; the equality test fails on 4.07-4.14) without this PR.

dra27 commented:
(obviously the test is irrelevant for 5.x as there's no `Stdlib.Stream` to test against)

avsm reviewed (approved):
This looks right to me now.  The only thing that become a problem is the use of `Jbuild_plugin` for the dune file, which is something of a slow spot for incremental dune builds and generally not recommended.  However, this is something we could fix up in a future iteration of camlp-streams rather than delay this PR (since it is needed to unblock `odoc-parser` and a plethora of OCaml 5.0 opam packages)

dra27 commented:
Argh - this won't work on MSVC (for 4.x) - fix incoming!

dra27 committed:
Fix MSVC for 4.07+

4.02-4.06 remain broken.

dra27 commented:
We hit the same problem as stdlib-shims that we MSVC doesn't support empty archives until OCaml 4.11.

Having battled with various things, I think the best thing for now is to accept that this library doesn't work properly on OCaml 4.02-4.06 for MSVC only. I've changed it so that OCaml 4.07-4.14 (i.e. all the versions of OCaml with the `Stdlib` module) do the same as 4.14 - they create a `Genlex` and `Stream` which are compatible with the Standard Library's. In many respects, that's more consistent.

I have a version which can do this with a static `dune` file, but it's not pretty - I have a solution for 4.02-4.06 for MSVC which works for installation, but not for Dune vendoring. On the basis that we do need a release, I'd propose this goes in now and the MSVC issue gets fixed in the future.

dra27 committed:
Empty library for OCaml 4.07 as well

We hit the issue corrected in PR#2041 in OCaml 4.08 - even in OCaml 4.07
with its prefixed Stdlib, we can't override modules opened from Stdlib.

nojb commented on dune:
```diff
@@ -0,0 +1,63 @@
+let () =
+  let stream = dune_fragment_for "Stream" "stream" in
+  let genlex = dune_fragment_for "Genlex" "genlex" in
+Printf.ksprintf send {|
```
Missing indent? :)

dra27 replied:
Fixed :)

nojb reviewed (approved):
LGTM!

dra27 commented:
@xavierleroy - please could we have a new release? I expect 5.0.1 is OK as a release number - we've not changed any "features"?

dra27 commented:
(I'm also happy to do the release, obviously)

xavierleroy commented:
Feel free to do a release when you think it's ready.  Thanks!
